ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java

```diff
@@ -0,0 +1,1240 @@
+import io.confluent.ksql.parser.tree.NullLiteral;
+import io.confluent.ksql.parser.tree.QualifiedNameReference;
+import io.confluent.ksql.parser.tree.Query;
+import io.confluent.ksql.parser.tree.QueryBody;
+import io.confluent.ksql.parser.tree.QuerySpecification;
+import io.confluent.ksql.parser.tree.Relation;
+import io.confluent.ksql.parser.tree.RenameColumn;
+import io.confluent.ksql.parser.tree.SampledRelation;
+import io.confluent.ksql.parser.tree.SearchedCaseExpression;
+import io.confluent.ksql.parser.tree.Select;
+import io.confluent.ksql.parser.tree.SelectItem;
+import io.confluent.ksql.parser.tree.SessionWindowExpression;
+import io.confluent.ksql.parser.tree.SetOperation;
+import io.confluent.ksql.parser.tree.SetSession;
+import io.confluent.ksql.parser.tree.ShowCatalogs;
+import io.confluent.ksql.parser.tree.ShowColumns;
+import io.confluent.ksql.parser.tree.ShowCreate;
+import io.confluent.ksql.parser.tree.ShowFunctions;
+import io.confluent.ksql.parser.tree.ShowPartitions;
+import io.confluent.ksql.parser.tree.SimpleCaseExpression;
+import io.confluent.ksql.parser.tree.SimpleGroupBy;
+import io.confluent.ksql.parser.tree.SingleColumn;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.parser.tree.Statements;
+import io.confluent.ksql.parser.tree.StringLiteral;
+import io.confluent.ksql.parser.tree.Struct;
+import io.confluent.ksql.parser.tree.SubqueryExpression;
+import io.confluent.ksql.parser.tree.SubscriptExpression;
+import io.confluent.ksql.parser.tree.SymbolReference;
+import io.confluent.ksql.parser.tree.Table;
+import io.confluent.ksql.parser.tree.TableElement;
+import io.confluent.ksql.parser.tree.TableSubquery;
+import io.confluent.ksql.parser.tree.TimeLiteral;
+import io.confluent.ksql.parser.tree.TimestampLiteral;
+import io.confluent.ksql.parser.tree.TumblingWindowExpression;
+import io.confluent.ksql.parser.tree.Values;
+import io.confluent.ksql.parser.tree.WhenClause;
+import io.confluent.ksql.parser.tree.Window;
+import io.confluent.ksql.parser.tree.WindowExpression;
+import io.confluent.ksql.parser.tree.WindowFrame;
+import io.confluent.ksql.parser.tree.WithQuery;
+import io.confluent.ksql.util.KsqlException;
+
+public class StatementRewriter extends DefaultAstVisitor<Node, Object> {
+
+  protected Node visitNode(Node node, Object context) {
+    return null;
+  }
+
+  protected Node visitExpression(Expression node, Object context) {
+    return node;
+  }
+
+  protected Node visitExtract(Extract node, Object context) {
+
+    if (node.getLocation().isPresent()) {
+      return new Extract(node.getLocation().get(),
+          (Expression) process(node.getExpression(), context),
+          node.getField());
+    } else {
+      return new Extract((Expression) process(node.getExpression(), context),
+          node.getField());
+    }
+
+  }
+
+  protected Node visitArithmeticBinary(ArithmeticBinaryExpression node, Object context) {
+    if (node.getLocation().isPresent()) {
```

Comment:
here and elsewhere could be:
```
return node.getLocation().
        map(location ->
            new ArithmeticBinaryExpression(location, node.getType(),
                (Expression) process(node.getLeft(), context),
                (Expression) process(node.getRight(), context)))
        .orElse(new ArithmeticBinaryExpression(node.getType(),
            (Expression) process(node.getLeft(), context),
            (Expression) process(node.getRight(), context)));
```

Reply:
That's an interesting way of writing `if/else`. Made your suggested changes.

Reply:
This is how `Optional` is intended to be used. :D